packages/protocol/scope-map-cache.ts

```diff
@@ -0,0 +1,24 @@
+import { Location } from "@replayio/protocol";
+
+import { ThreadFront } from "./thread/thread";
+import { defer } from "./utils";
+
+export default class ScopeMapCache {
+  private cache = new Map<string, Promise<Record<string, string>>>();
+
+  async getScopeMap(location: Location): Promise<Record<string, string>> {
+    const cacheKey = `${location.sourceId}:${location.line}:${location.column}`;
+
+    if (this.cache.has(cacheKey)) {
+      return this.cache.get(cacheKey)!;
+    }
+
+    const { promise, resolve } = defer<Record<string, string>>();
+    this.cache.set(cacheKey, promise);
+    const { client } = await import("./socket");
+    const { map } = await client.Debugger.getScopeMap({ location }, ThreadFront.sessionId!);
+    const scopeMap: Record<string, string> = map ? Object.fromEntries(map) : {};
```

Comment:
Is `Record<string, string>` the correct type here? I would have expected it to be something like `{[key: string]: string}` or `Record<string, string | undefined>`?

Reply:
I also thought this, but it's not! `Record<A, B>` says "For every A I will have you back a B. I will never hand back a null or undefined, only a B". `{[key: string]: string}` says "I'm an object that only contains string keys and string values, but I don't necessarily have an entry for every possible key string, sometimes I will hand back `undefined`". At least, I'm 95% sure it works this way, since it bit me on a previous side project 😄